util/src/main/java/com/coderanch/util/cli/InputUtility.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (C) 2018 Coderanch.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.coderanch.util.cli;
+
+import java.io.*;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+
+/**
+ * Multipurpose input utility for use with a command line interface.
+ * The user can pass a prompt and optional validation.
+ * The utility will read the provided input stream and return a possibly validated String or primitive.
+ */
+public final class InputUtility implements Closeable, AutoCloseable {
+
+    /**
+     * Possible inputs for yes.
+     */
+    private static final List<String> YES = Collections.unmodifiableList(List.of("y", "yes"));
+
+    /**
+     * Possible inputs for no.
+     */
+    private static final List<String> NO = Collections.unmodifiableList(List.of("n", "no"));
+
+    /**
+     * The reader to read user responses with.
+     */
+    private final BufferedReader reader;
+
+    /**
+     * If no input stream is provided, System.in will be used.
+     */
+    public InputUtility() {
+        this(System.in, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Constructor that takes an input stream to be used.
+     *
+     * @param inputStream input stream to be used.
+     * @param charset     charset to be used.
+     */
+    public InputUtility(InputStream inputStream, Charset charset) {
+        reader = new BufferedReader(new InputStreamReader(inputStream, charset));
+    }
+
+    /**
+     * Displays the prompt and reads the input stream, returning a possibly validated String.
+     *
+     * @param prompt          the prompt to display to the user.
+     * @param stringPredicate the predicate to use for validation.
+     * @return a possibly validated String.
+     * @throws IOException when there's a problem with {@link InputStream}
+     */
+    public String nextString(String prompt, Predicate<? super String> stringPredicate) throws IOException {
+        System.out.println(prompt);
+        while (true) {
+            var line = reader.readLine();
+            if (stringPredicate.test(line)) {
+                return line;
+            }
+            else {
+                System.out.println("Invalid input.");
+            }
+        }
```

Comment:
Don't use `true` and `false` as loop conditions.
```suggestion
        String line;
        
        for (
            line = reader.readLine();
            !stringPredicate.test(line);
            line = reader.readLine();
        ) {
            System.out.println("Invalid input.");
        }
        
        return line;
```